### ISSUE TITLE
Reconsider signatures of Set#has and Array#includes to allow search value with overlapping type

### ISSUE THREAD

**Issue by Ginden** [microsoft/TypeScript#45472]
# lib Update Request

## Configuration Check

My compilation *target* is `ES2020` and my *lib* is `the default`.

## Missing / Incorrect Definition

Let's talk about `Set#has()` and `Array#includes()` and consider following code:

## Sample Code

Define example enum
```typescript
enum RuleType {
   AT_SPECIFIC_HOUR = 'AT_SPECIFIC_HOUR',
   AFTER_EXPECTED_ARRIVAL = 'AFTER_EXPECTED_ARRIVAL',
   ON_COMPLAINT_RECEIVED = 'ON_COMPLAINT_RECEIVED',
}
```

```typescript
const timeBasedRules = new Set([RuleType.AT_SPECIFIC_HOUR, RuleType.AFTER_EXPECTED_ARRIVAL]);
const testedValue = RuleType.ON_COMPLAINT_RECEIVED as RuleType;
timeBasedRules.has(testedValue); // Error
```

Also:
```typescript
const timeBasedRules = [RuleType.AT_SPECIFIC_HOUR, RuleType.AFTER_EXPECTED_ARRIVAL] as const; // Interestingly, Set is inferred as AT_SPECIFIC_HOUR | AFTER_EXPECTED_ARRIVAL, but array is inferred as RuleType[]
const testedValue = RuleType.ON_COMPLAINT_RECEIVED as RuleType;

timeBasedRules.includes(testedValue); // error
```

Why do I consider this behaviour counterintuitive? Because **provably equivalent code compiles without error** [1]:

```typescript
const timeBasedRules = [RuleType.AT_SPECIFIC_HOUR, RuleType.AFTER_EXPECTED_ARRIVAL] as const;
const testedValue = RuleType.ON_COMPLAINT_RECEIVED as RuleType;

timeBasedRules.some(v => v === testedValue)

```

**tl;dr: if `V` and `T` has overlap, then `([] as T[]).includes(null as uknown as V)`  should be allowed**.

## Documentation Link

> includes compares searchElement to the elements of the array, in ascending order, **using the SameValueZero algorithm**

[23.1.3.13 Array.prototype.includes ( searchElement [ , fromIndex ] )](https://tc39.es/ecma262/multipage/indexed-collections.html#sec-array.prototype.includes)

[1] There are 2 edge cases for using `IsStrictlyEqual` instead of `SameValueZero`, but these are not representable in TypeScript anyway.

**Comment by MartinJohns** [comment]
Duplicate of #26255 / #14520.

**Comment by Ginden** [comment]
@MartinJohns You are right, thanks.
